relays/substrate/src/messages_source.rs

```diff
@@ -0,0 +1,221 @@
+	}
+
+	async fn latest_confirmed_received_nonce(
+		&self,
+		id: SourceHeaderIdOf<P>,
+	) -> Result<(SourceHeaderIdOf<P>, P::MessageNonce), Self::Error> {
+		let encoded_response = self
+			.client
+			.state_call(
+				// TODO: https://github.com/paritytech/parity-bridges-common/issues/457
+				"OutboundLaneApi_latest_received_nonce".into(),
+				Bytes(self.lane.encode()),
+				Some(id.1),
+			)
+			.await?;
+		let latest_received_nonce: P::MessageNonce =
+			Decode::decode(&mut &encoded_response.0[..]).map_err(SubstrateError::ResponseParseFailed)?;
+		Ok((id, latest_received_nonce))
+	}
+
+	async fn prove_messages(
+		&self,
+		id: SourceHeaderIdOf<P>,
+		nonces: RangeInclusive<P::MessageNonce>,
+		include_outbound_lane_state: bool,
+	) -> Result<(SourceHeaderIdOf<P>, RangeInclusive<P::MessageNonce>, P::MessagesProof), Self::Error> {
+		let (weight, proof) = self
+			.client
+			.prove_messages(
+				self.instance,
+				self.lane,
+				nonces.clone(),
+				include_outbound_lane_state,
+				id.1,
+			)
+			.await?;
+		let proof = (id.1, proof, self.lane, *nonces.start(), *nonces.end());
+		Ok((id, nonces, (weight, proof)))
+	}
+
+	async fn submit_messages_receiving_proof(
+		&self,
+		generated_at_block: TargetHeaderIdOf<P>,
+		proof: P::MessagesReceivingProof,
+	) -> Result<(), Self::Error> {
+		let tx = self
+			.tx_maker
+			.make_messages_receiving_proof_transaction(generated_at_block, proof)
+			.await?;
+		self.client.submit_extrinsic(Bytes(tx.encode())).await?;
+		Ok(())
+	}
+}
+
+/// Read client state.
+pub async fn read_client_state<SelfChain, BridgedHeaderHash, BridgedHeaderNumber>(
+	self_client: &Client<SelfChain>,
+	bridged_chain_name: &str,
+) -> Result<ClientState<HeaderIdOf<SelfChain>, HeaderId<BridgedHeaderHash, BridgedHeaderNumber>>, SubstrateError>
+where
+	SelfChain: Chain,
+	SelfChain::Header: DeserializeOwned,
+	SelfChain::Index: DeserializeOwned,
+	BridgedHeaderHash: Decode,
+	BridgedHeaderNumber: Decode,
+{
+	// let's read our state first: we need best finalized header hash on **this** chain
+	let self_best_finalized_header_hash = self_client.best_finalized_header_hash().await?;
+	let self_best_finalized_header = self_client.header_by_hash(self_best_finalized_header_hash).await?;
+	let self_best_finalized_id = HeaderId(*self_best_finalized_header.number(), self_best_finalized_header_hash);
+
+	// now let's read id of best finalized peer header at our best finalized block
+	let best_finalized_peer_on_self_method = format!("{}HeaderApi_finalized_block", bridged_chain_name);
+	let encoded_best_finalized_peer_on_self = self_client
+		.state_call(
+			best_finalized_peer_on_self_method,
+			Bytes(Vec::new()),
+			Some(self_best_finalized_header_hash),
+		)
+		.await?;
+	let decoded_best_finalized_peer_on_self: (BridgedHeaderNumber, BridgedHeaderHash) =
+		Decode::decode(&mut &encoded_best_finalized_peer_on_self.0[..]).map_err(SubstrateError::ResponseParseFailed)?;
+	let peer_on_self_best_finalized_id = HeaderId(
```

Comment:
Maybe we can add a `From<(Num, Hash)>` implementation for `HeaderId` so this can just be an `.into()` call

Reply:
Why do you think `.into()` is better than `HeaderId()`? The `HeaderId` is here, because tuples were not intuitive - you need some context to understand what's this tuple is. Going back to `(...).into()` doesn't help imo

Reply:
It seems a little silly to me that we're breaking up this tuple (`decoded_best_finalized_peer_on_self`) just to turn it back into a `HeaderId`

Reply:
I'd rather be for `HeaderId::from_tuple(...)` instead of `.into()`